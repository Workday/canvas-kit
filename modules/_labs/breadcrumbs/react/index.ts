import Breadcrumbs from './lib/Breadcrumbs';

export default Breadcrumbs;
export {Breadcrumbs};
export * from './lib/Breadcrumbs';
