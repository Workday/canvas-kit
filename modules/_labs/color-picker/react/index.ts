export {default, default as ColorPicker, ColorPickerProps} from './lib/ColorPicker';
export {Swatch} from './lib/parts/Swatch';
