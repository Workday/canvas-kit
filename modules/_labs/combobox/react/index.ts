import Combobox from './lib/Combobox';

export default Combobox;
export {Combobox};
export * from './lib/Combobox';
