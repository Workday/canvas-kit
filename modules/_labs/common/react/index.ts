export * from './lib/Flex';
export * from './lib/Box';
