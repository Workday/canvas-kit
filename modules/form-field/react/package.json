{
  "name": "@workday/canvas-kit-react-form-field",
  "version": "3.1.0",
  "description": "Form field component to wrap inputs",
  "author": "Workday, Inc. (https://www.workday.com)",
  "license": "Apache-2.0",
  "main": "dist/commonjs/index.js",
  "module": "dist/es6/index.js",
  "sideEffects": false,
  "types": "dist/es6/index.d.ts",
  "repository": {
    "type": "git",
    "url": "http://github.com/Workday/canvas-kit/tree/master/modules/form-field/react"
  },
  "files": [
    "dist/",
    "lib/",
    "index.ts"
  ],
  "scripts": {
    "watch": "yarn build:es6 -w",
    "test": "echo \"Error: no test specified\" && exit 1",
    "clean": "rimraf dist && rimraf .build-info && mkdirp dist",
    "build:cjs": "tsc -p tsconfig.cjs.json",
    "build:es6": "tsc -p tsconfig.es6.json",
    "build:rebuild": "npm-run-all clean build",
    "build": "npm-run-all --parallel build:cjs build:es6"
  },
  "keywords": [
    "canvas",
    "canvas-kit",
    "react",
    "components",
    "workday",
    "input-common"
  ],
  "peerDependencies": {
    "react": ">= 16.8 < 17"
  },
  "dependencies": {
    "@emotion/core": "^10.0.21",
    "@emotion/styled": "^10.0.17",
    "@workday/canvas-kit-react-common": "^3.1.0",
    "@workday/canvas-kit-react-core": "^3.1.0",
    "uuid": "^3.3.3"
  },
  "devDependencies": {
    "@workday/canvas-kit-react-checkbox": "^3.1.0",
    "@workday/canvas-kit-react-radio": "^3.1.0",
    "@workday/canvas-kit-react-select": "^3.1.0",
    "@workday/canvas-kit-react-text-input": "^3.1.0"
  }
}
