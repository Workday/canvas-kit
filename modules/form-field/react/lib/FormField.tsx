import * as React from 'react';
import styled from '@emotion/styled';
import {spacing} from '@workday/canvas-kit-react-core';
import {GrowthBehavior, ErrorType} from '@workday/canvas-kit-react-common';
import Hint from './Hint';
import Label from './Label';
import {FormFieldLabelPosition, FormFieldLabelPositionBehavior} from './types';
import uuid from 'uuid/v4';

export interface FormFieldProps extends React.HTMLAttributes<HTMLDivElement>, GrowthBehavior {
  labelPosition: FormFieldLabelPosition;
  label?: React.ReactNode;
  hintText?: React.ReactNode;
  hintId?: string;
  inputId?: string;
  error?: ErrorType;
  required?: boolean;
  useFieldset: boolean;
  children: React.ReactNode;
}

export interface FormFieldErrorBehavior {
  error?: ErrorType;
}

// Use a fieldset element for accessible radio groups
const FormFieldFieldsetContainer = styled('fieldset')<FormFieldLabelPositionBehavior>({
  padding: 0,
  margin: 0,
  border: 0,
});

const FormFieldContainer = styled('div')<FormFieldLabelPositionBehavior>(({labelPosition}) => {
  if (labelPosition === FormFieldLabelPosition.Left) {
    return {
      display: 'flex',
      marginBottom: spacing.m,
    };
  }

  return {
    display: 'block',
    marginBottom: spacing.m,
  };
});

const FormFieldInputContainer = styled('div')<GrowthBehavior & FormFieldLabelPositionBehavior>(
  ({grow, labelPosition}) => {
    if (grow) {
      if (
        labelPosition === FormFieldLabelPosition.Left ||
        labelPosition === FormFieldLabelPosition.Hidden
      ) {
        return {
          flexGrow: 1,
        };
      }

      return {
        display: 'block',
      };
    }

    if (labelPosition === FormFieldLabelPosition.Left) {
      return {
        display: 'inline-flex',
        alignItems: 'center',
        flexWrap: 'wrap',
      };
    }

    return {
      display: 'inline-block',
    };
  }
);

export default class FormField extends React.Component<FormFieldProps> {
  static LabelPosition = FormFieldLabelPosition;
  static ErrorType = ErrorType;

  static defaultProps = {
    labelPosition: FormField.LabelPosition.Top,
    useFieldset: false,
  };

  private inputId: string = this.props.inputId || uuid();

  private renderChildren = (child: React.ReactElement): React.ReactNode => {
    if (React.isValidElement<any>(child)) {
      const props: GrowthBehavior &
        FormFieldErrorBehavior &
        React.InputHTMLAttributes<HTMLInputElement> = {
        ...child.props,
      };

      if (this.props.grow) {
        props.grow = this.props.grow;
      }

      if (typeof this.props.error !== 'undefined') {
        props.error = this.props.error;

        if (this.props.hintId) {
          props['aria-describedby'] = this.props.hintId;
        }

        if (this.props.error === ErrorType.Error) {
          props['aria-invalid'] = true;
        }
      }

      props.id = this.inputId;

      return React.cloneElement(child, props);
    }

    return child;
  };

  render() {
    const {
      label,
      hintText,
      hintId,
      inputId,
      grow,
      children,
      useFieldset,
      labelPosition,
      error,
      required,
      ...elemProps
    } = this.props;

    const field = (
      <FormFieldContainer labelPosition={labelPosition} {...elemProps}>
        {typeof label === 'string' ? (
          <Label
            labelPosition={labelPosition}
            htmlFor={this.inputId}
            isLegend={useFieldset}
            required={!!required}
            accessibleHide={labelPosition === FormFieldLabelPosition.Hidden}
          >
            {label}
          </Label>
        ) : (
          label
        )}

        <FormFieldInputContainer grow={grow} labelPosition={labelPosition}>
          {React.Children.map(children, this.renderChildren)}
          {hintText && (
            <Hint error={error} id={hintId}>
              {hintText}
            </Hint>
          )}
        </FormFieldInputContainer>
      </FormFieldContainer>
    );

    return useFieldset ? <FormFieldFieldsetContainer>{field}</FormFieldFieldsetContainer> : field;
  }
}
