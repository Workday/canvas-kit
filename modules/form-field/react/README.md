# Canvas Kit Form Field

Form field wrapper component that makes inputs accessible. It also includes complementary elements
for form inputs (labels, errors, etc.).

## Installation

```sh
yarn add @workday/canvas-kit-react
```

or

```sh
yarn add @workday/canvas-kit-react-form-field
```

# FormField

## Usage

```tsx
import * as React from 'react;
import FormField from '@workday/canvas-kit-react-form-field';
import TextInput from '@workday/canvas-kit-react-text-input';

<FormField
  label="Label"
  inputId="my-input-field"
  error={FormField.ErrorType.Error}
  hintText="This field is required."
  hintId="my-input-field-error"
  required={true}
>
  <TextInput value={this.state.value} onChange={() => {}}/>
</FormField>
```

## Static Properties

#### `ErrorType: ErrorType`

```tsx
<TextInput error={TextInput.ErrorType.Alert} />
```

---

#### `LabelPosition: FormFieldLabelPosition`

```tsx
<TextInput labelPosition={TextInput.LabelPosition.Left} />
```

## Component Props

### Required

#### `children: React.ReactNode`

> The input component to wrap.

### Optional

#### `error: ErrorType`

> The type of error to display, if any. This prop will be passed to the input component if
> applicable.

| Type  | Description                     |
| ----- | ------------------------------- |
| Error | Red outline with error icon.    |
| Alert | Yellow outline with alert icon. |

Default: `undefined`

---

#### `grow: boolean`

> Whether or not the field will expand to the width of the container. This prop will be passed to
> the input component if applicable.

Default: `false`

---

#### `hintText: React.ReactNode`

> The message displayed below the input field. Required if `error` is defined.

Default: `undefined`

---

#### `hintId: string`

> The ID of message displayed below the input field. Required for accessibile aria-definedby
> attribute. Required if `error` and `hintText` are defined.

---

#### `label: string`

> The label text displayed for the input.

---

#### `requiredlabel: string`

> The required label text displayed for the required input.

---

#### `inputId: string`

> The ID of the input child. If an ID is not specified on the input child, this will be used as it's
> ID. Used for label's `htmlFor` attribute. This is required for accessiblity if `label` is defined.

---

#### `labelPosition: FormFieldLabelPosition`

> The position of the label relative to the input field.

| Type | Description                                   |
| ---- | --------------------------------------------- |
| Top  | Display label on the left, inline with input. |
| Left | Display label blocked on top of input         |

Default: `LabelPosition.Top`

---

#### `required: boolean`

> It set, the label of the field will be suffixed by a red astrisk.

Default: `false`

---

#### `useFieldset: boolean`

> If true, FormField uses a `fieldset` and a `legend` element instead of a div and a label. This is
> required for accessibility on radio groups. If you're using a `RadioGroup` inside of FormField,
> make sure you set this to true.

Default: `false`

---

# Hint

> Component implementing hint text styling for forms.

## Usage

```tsx
import * as React from 'react;
import {Hint} from '@workday/canvas-kit-react-form-field';

<Hint error={Hint.ErrorType.Error}>
  This field is required.
</Hint>
```

## Static Properties

#### `ErrorType: ErrorType`

```tsx
<Hint error={Hint.ErrorType.Alert} />
```

## Component Props

### Optional

#### `error: ErrorType`

> The type of error to display, if any.

| Type  | Description                     |
| ----- | ------------------------------- |
| Error | Displays "Error:" prefix        |
| Alert | Yellow outline with alert icon. |

Default: `undefined`

---

# Label

> Component implementing label text styling for forms.

## Usage

```tsx
import * as React from 'react;
import {Label} from '@workday/canvas-kit-react-form-field';

<Label error={Label.LabelPosition.Left}>
  Label
</Label>
```

## Static Properties

#### `Position: FormFieldLabelPosition`

```tsx
<Label error={Label.Position.Left} />
```

## Component Props

### Optional

#### `labelPosition: FormFieldLabelPosition`

> Position of the label.

| Type | Description                                   |
| ---- | --------------------------------------------- |
| Top  | Display label on the left, inline with input. |
| Left | Display label blocked on top of input         |

Default: `LabelPosition.Top`

---

#### `htmlFor: string`

> The ID of the input being labelled. This is required for accessiblity.
