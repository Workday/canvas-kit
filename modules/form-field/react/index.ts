import FormField from './lib/FormField';
import Hint from './lib/Hint';
import Label from './lib/Label';
import InputIconContainer from './lib/InputIconContainer';

export default FormField;
export {FormField, Hint, Label, InputIconContainer};
export * from './lib/types';
export * from './lib/FormField';
export * from './lib/Hint';
export * from './lib/Label';
export * from './lib/InputIconContainer';
