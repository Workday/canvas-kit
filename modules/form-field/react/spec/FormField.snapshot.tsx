import * as React from 'react';
import * as renderer from 'react-test-renderer';
import FormField from '../lib/FormField';

describe('FormField Snapshots', () => {
  test('renders as expected', () => {
    const component = renderer.create(
      <FormField>
        <input type="text" />
      </FormField>
    );
    expect(component).toMatchSnapshot();
  });

  test('left label', () => {
    const component = renderer.create(
      <FormField labelPosition={FormField.LabelPosition.Left}>
        <input type="text" />
      </FormField>
    );
    expect(component).toMatchSnapshot();
  });

  test('grow', () => {
    const component = renderer.create(
      <FormField grow={true}>
        <input type="text" />
      </FormField>
    );
    expect(component).toMatchSnapshot();
  });

  test('grow with left label', () => {
    const component = renderer.create(
      <FormField labelPosition={FormField.LabelPosition.Left} grow={true}>
        <input type="text" />
      </FormField>
    );
    expect(component).toMatchSnapshot();
  });
});
