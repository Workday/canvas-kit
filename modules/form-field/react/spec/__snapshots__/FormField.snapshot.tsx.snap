// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField Snapshots grow 1`] = `
.emotion-2 {
  display: block;
  margin-bottom: 24px;
}

.emotion-0 {
  display: block;
}

<div
  className="emotion-2 emotion-3"
>
  <div
    className="emotion-0 emotion-1"
  >
    <input
      grow={true}
      type="text"
    />
  </div>
</div>
`;

exports[`FormField Snapshots grow with left label 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 24px;
}

.emotion-0 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div
  className="emotion-2 emotion-3"
>
  <div
    className="emotion-0 emotion-1"
  >
    <input
      grow={true}
      type="text"
    />
  </div>
</div>
`;

exports[`FormField Snapshots left label 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 24px;
}

.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

<div
  className="emotion-2 emotion-3"
>
  <div
    className="emotion-0 emotion-1"
  >
    <input
      type="text"
    />
  </div>
</div>
`;

exports[`FormField Snapshots renders as expected 1`] = `
.emotion-2 {
  display: block;
  margin-bottom: 24px;
}

.emotion-0 {
  display: inline-block;
}

<div
  className="emotion-2 emotion-3"
>
  <div
    className="emotion-0 emotion-1"
  >
    <input
      type="text"
    />
  </div>
</div>
`;

exports[`FormField Snapshots renders required marking 1`] = `
.emotion-2 {
  display: block;
  margin-bottom: 24px;
}

.emotion-0 {
  display: inline-block;
}

<div
  className="emotion-2 emotion-3"
>
  <div
    className="emotion-0 emotion-1"
  >
    <input
      type="text"
    />
  </div>
</div>
`;
